File: packages/vault-memory/src/awakeReplay.ts
import { Database } from "bun:sqlite";
import { Memory, bumpReplayCount, storeReplay, parseMemoryRow } from "./db.js";

export class AwakeReplayEngine {
  constructor(private db: Database, private vaultPath: string) {}

  /**
   * Reverse Replay (Credit Assignment)
   * Triggered by: job completion, task completion, [DONE:] tag
   */
  async reverseReplay(opts: {
    triggerRef: string;
    triggerSource: string;
    entities?: string[];
    timeWindowMs?: number; // default: 24h
  }): Promise<{ replayedCount: number; creditDelta: number }> {
    const timeWindow = opts.timeWindowMs ?? 24 * 60 * 60 * 1000;
    const cutoff = new Date(Date.now() - timeWindow).toISOString();

    // 1. Find memories created during the job's lifetime or related by entities
    // We look for direct job memories and associative memories
    const directJobPattern = `%${opts.triggerRef}%`;
    
    let query = `
      SELECT * FROM memories 
      WHERE (source LIKE ? OR summary LIKE ?)
      AND created_at > ?
    `;
    const params: any[] = [directJobPattern, directJobPattern, cutoff];

    if (opts.entities && opts.entities.length > 0) {
      const entityConditions = opts.entities.map(() => "entities LIKE ?").join(" OR ");
      query += ` OR (${entityConditions})`;
      opts.entities.forEach(e => params.push(`%${e}%`));
    }

    const memories = this.db.prepare(query).all(...params).map(parseMemoryRow) as Memory[];
    
    if (memories.length === 0) {
      return { replayedCount: 0, creditDelta: 0 };
    }

    // 2. Order chronologically (ASC) to form the sequence
    const sequence = memories.sort((a, b) => a.created_at.localeCompare(b.created_at));
    const memoryIds = sequence.map(m => m.id);

    // 3. Apply credit delta: base 0.05 + 0.01 per chain step (capped at 0.15)
    const creditDelta = Math.min(0.15, 0.05 + (memories.length * 0.01));

    // Update importance in DB
    const placeholders = memoryIds.map(() => "?").join(",");
    this.db.prepare(`
      UPDATE memories 
      SET importance = MIN(1.0, importance + ?) 
      WHERE id IN (${placeholders})
    `).run(creditDelta, ...memoryIds);

    // 4. Bump replay_count on all involved memories (bookmarking)
    bumpReplayCount(this.db, memoryIds);

    // 5. Store replay record
    storeReplay(this.db, {
      trigger_type: 'reverse',
      trigger_source: opts.triggerSource,
      trigger_ref: opts.triggerRef,
      memory_ids: memoryIds,
      sequence: memoryIds,
      credit_delta: creditDelta
    });

    return { replayedCount: memories.length, creditDelta };
  }

  /**
   * Forward Replay (Planning/Preplay)
   * Triggered by: new job creation, new task creation
   */
  async forwardReplay(opts: {
    triggerRef: string;
    triggerSource: string;
    instructionText: string;
    limit?: number; // default: 5
  }): Promise<{ precedents: Memory[]; replayedCount: number }> {
    // 1. Extract entities/topics from instruction text via fast regex
    const cues = this.extractCues(opts.instructionText);
    
    // 2. Query memories sharing entities or topics
    const precedents = this.findRelatedMemories(cues, { limit: opts.limit ?? 5 });

    if (precedents.length === 0) {
      return { precedents: [], replayedCount: 0 };
    }

    const memoryIds = precedents.map(m => m.id);

    // 3. Bump replay_count (bookmarking)
    bumpReplayCount(this.db, memoryIds);

    // 4. Store replay record
    storeReplay(this.db, {
      trigger_type: 'forward',
      trigger_source: opts.triggerSource,
      trigger_ref: opts.triggerRef,
      memory_ids: memoryIds,
      sequence: memoryIds, // Sequence is same as IDs for forward replay (ranked by importance)
      credit_delta: 0
    });

    return { precedents, replayedCount: precedents.length };
  }

  /**
   * Fast cue extraction (regex only, <50ms)
   */
  private extractCues(text: string): { entities: string[]; topics: string[] } {
    const entities = new Set<string>();
    const topics = new Set<string>();

    // Capitalized phrases (Entities)
    const capMatches = text.matchAll(/\b([A-Z][a-zA-Z]+(?: [A-Z][a-zA-Z]+)*)\b/g);
    for (const match of capMatches) {
      if (match[1].length > 3) entities.add(match[1]);
    }

    // #tags (Topics)
    const tagMatches = text.matchAll(/#([a-zA-Z0-9_]+)/g);
    for (const match of tagMatches) {
      topics.add(match[1]);
    }

    // @mentions (Entities)
    const mentionMatches = text.matchAll(/@([a-zA-Z0-9_]+)/g);
    for (const match of mentionMatches) {
      entities.add(match[1]);
    }

    // Quoted strings (Entities/Specifics)
    const quoteMatches = text.matchAll(/"([^"]+)"|'([^']+)'/g);
    for (const match of quoteMatches) {
      const val = match[1] || match[2];
      if (val && val.length > 2) entities.add(val);
    }

    return {
      entities: Array.from(entities),
      topics: Array.from(topics)
    };
  }

  /**
   * Memory retrieval by cue matching
   */
  private findRelatedMemories(cues: { entities: string[]; topics: string[] }, opts: { limit: number }): Memory[] {
    if (cues.entities.length === 0 && cues.topics.length === 0) {
      // Fallback: just return the most important recent memories
      return this.db.prepare(`
        SELECT * FROM memories 
        WHERE consolidated = 0 
        ORDER BY importance DESC, created_at DESC 
        LIMIT ?
      `).all(opts.limit).map(parseMemoryRow) as Memory[];
    }

    const conditions: string[] = [];
    const params: any[] = [];

    cues.entities.forEach(e => {
      conditions.push("entities LIKE ?");
      params.push(`%${e}%`);
    });

    cues.topics.forEach(t => {
      conditions.push("topics LIKE ?");
      params.push(`%${t}%`);
    });

    const query = `
      SELECT * FROM memories 
      WHERE (${conditions.join(" OR ")})
      ORDER BY importance DESC, created_at DESC 
      LIMIT ?
    `;
    params.push(opts.limit);

    return this.db.prepare(query).all(...params).map(parseMemoryRow) as Memory[];
  }
}

File: crates/hq-memory/src/awake_replay.rs
//! AwakeReplayEngine — reverse (credit assignment) and forward (planning) replay.
//!
//! Ported from vault-memory/src/awakeReplay.ts.
//!
//! Reverse Replay: triggered by job/task completion. Finds memories created during
//! the job's lifetime and boosts their importance (credit assignment).
//!
//! Forward Replay: triggered by new job/task creation. Finds related precedent
//! memories by cue matching (entities, topics).

use anyhow::Result;
use hq_db::Database;
use regex::Regex;
use std::collections::HashSet;
use std::path::PathBuf;
use std::sync::LazyLock;
use tracing::info;

use crate::db::{bump_replay_count, store_replay};
use crate::types::{Memory, ReplayTriggerType};

/// Awake replay engine for credit assignment and planning.
pub struct AwakeReplayEngine {
    db: Database,
    #[allow(dead_code)]
    vault_path: PathBuf,
}

/// Result of a reverse replay operation.
#[derive(Debug, Clone)]
pub struct ReverseReplayResult {
    pub replayed_count: usize,
    pub credit_delta: f64,
}

/// Result of a forward replay operation.
#[derive(Debug, Clone)]
pub struct ForwardReplayResult {
    pub precedents: Vec<Memory>,
    pub replayed_count: usize,
}

/// Cues extracted from text for memory retrieval.
struct Cues {
    entities: Vec<String>,
    topics: Vec<String>,
}

// Regex patterns for cue extraction
static CAP_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b([A-Z][a-zA-Z]+(?:\s[A-Z][a-zA-Z]+)*)\b").unwrap()
});

static TAG_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"#([a-zA-Z0-9_]+)").unwrap()
});

static MENTION_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"@([a-zA-Z0-9_]+)").unwrap()
});

static QUOTE_PATTERN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#""([^"]+)"|'([^']+)'"#).unwrap()
});

impl AwakeReplayEngine {
    pub fn new(db: Database, vault_path: PathBuf) -> Self {
        Self { db, vault_path }
    }

    /// Reverse Replay (Credit Assignment)
    /// Triggered by: job completion, task completion, [DONE:] tag.
    ///
    /// Finds memories created during the job's lifetime or related by entities,
    /// boosts their importance, bumps replay_count, stores a replay record.
    pub fn reverse_replay(
        &self,
        trigger_ref: &str,
        trigger_source: &str,
        entities: Option<&[String]>,
        time_window_ms: Option<i64>,
    ) -> Result<ReverseReplayResult> {
        let time_window = time_window_ms.unwrap_or(24 * 60 * 60 * 1000);
        let cutoff = chrono::Utc::now() - chrono::Duration::milliseconds(time_window);
        let cutoff_iso = cutoff.to_rfc3339();

        let direct_pattern = format!("%{trigger_ref}%");

        // Build query for direct job memories + entity matches
        let memories = self.db.with_conn(|conn| {
            let mut sql = String::from(
                "SELECT * FROM memories WHERE (source LIKE ?1 OR summary LIKE ?2) AND created_at > ?3"
            );
            let mut params: Vec<Box<dyn rusqlite::types::ToSql>> = vec![
                Box::new(direct_pattern.clone()),
                Box::new(direct_pattern.clone()),
                Box::new(cutoff_iso.clone()),
            ];

            if let Some(ents) = entities {
                if !ents.is_empty() {
                    let entity_conditions: Vec<String> = ents.iter()
                        .map(|_| "entities LIKE ?".to_string())
                        .collect();
                    sql.push_str(&format!(" OR ({})", entity_conditions.join(" OR ")));
                    for e in ents {
                        params.push(Box::new(format!("%{e}%")));
                    }
                }
            }

            let mut stmt = conn.prepare(&sql)?;
            let param_refs: Vec<&dyn rusqlite::types::ToSql> = params.iter().map(|p| p.as_ref()).collect();
            let rows = stmt.query_map(param_refs.as_slice(), crate::db::parse_memory_row)?;

            let mut result = Vec::new();
            for row in rows {
                result.push(row?);
            }
            Ok(result)
        })?;

        if memories.is_empty() {
            return Ok(ReverseReplayResult { replayed_count: 0, credit_delta: 0.0 });
        }

        // Order chronologically (ASC)
        let mut sequence = memories;
        sequence.sort_by(|a, b| a.created_at.cmp(&b.created_at));
        let memory_ids: Vec<i64> = sequence.iter().map(|m| m.id).collect();

        // Credit delta: base 0.05 + 0.01 per chain step (capped at 0.15)
        let credit_delta = (0.05 + (sequence.len() as f64 * 0.01)).min(0.15);

        // Update importance in DB
        self.db.with_conn(|conn| {
            let placeholders: String = memory_ids.iter().map(|_| "?").collect::<Vec<_>>().join(",");
            let sql = format!(
                "UPDATE memories SET importance = MIN(1.0, importance + ?1) WHERE id IN ({placeholders})"
            );
            let mut stmt = conn.prepare(&sql)?;
            let mut params: Vec<Box<dyn rusqlite::types::ToSql>> = vec![Box::new(credit_delta)];
            for id in &memory_ids {
                params.push(Box::new(*id));
            }
            let param_refs: Vec<&dyn rusqlite::types::ToSql> = params.iter().map(|p| p.as_ref()).collect();
            stmt.execute(param_refs.as_slice())?;
            Ok(())
        })?;

        // Bump replay_count
        bump_replay_count(&self.db, &memory_ids)?;

        // Store replay record
        store_replay(
            &self.db,
            &ReplayTriggerType::Reverse,
            trigger_source,
            trigger_ref,
            &memory_ids,
            &memory_ids,
            credit_delta,
        )?;

        let replayed_count = sequence.len();
        info!(replayed_count, credit_delta, trigger_ref, "Reverse replay complete");

        Ok(ReverseReplayResult { replayed_count, credit_delta })
    }

    /// Forward Replay (Planning/Preplay)
    /// Triggered by: new job creation, new task creation.
    ///
    /// Extracts cues from instruction text, finds related precedent memories,
    /// bumps replay_count, stores a replay record.
    pub fn forward_replay(
        &self,
        trigger_ref: &str,
        trigger_source: &str,
        instruction_text: &str,
        limit: Option<usize>,
    ) -> Result<ForwardReplayResult> {
        let limit = limit.unwrap_or(5);

        // Extract cues from instruction text
        let cues = extract_cues(instruction_text);

        // Find related memories
        let precedents = self.find_related_memories(&cues, limit)?;

        if precedents.is_empty() {
            return Ok(ForwardReplayResult { precedents: Vec::new(), replayed_count: 0 });
        }

        let memory_ids: Vec<i64> = precedents.iter().map(|m| m.id).collect();

        // Bump replay_count
        bump_replay_count(&self.db, &memory_ids)?;

        // Store replay record
        store_replay(
            &self.db,
            &ReplayTriggerType::Forward,
            trigger_source,
            trigger_ref,
            &memory_ids,
            &memory_ids, // sequence same as IDs for forward replay
            0.0,
        )?;

        let replayed_count = precedents.len();
        info!(replayed_count, trigger_ref, "Forward replay complete");

        Ok(ForwardReplayResult { precedents, replayed_count })
    }

    /// Memory retrieval by cue matching.
    fn find_related_memories(&self, cues: &Cues, limit: usize) -> Result<Vec<Memory>> {
        self.db.with_conn(|conn| {
            if cues.entities.is_empty() && cues.topics.is_empty() {
                // Fallback: just return the most important recent memories
                let mut stmt = conn.prepare(
                    "SELECT * FROM memories WHERE consolidated = 0 ORDER BY importance DESC, created_at DESC LIMIT ?1"
                )?;
                let rows = stmt.query_map([limit as i64], crate::db::parse_memory_row)?;
                let mut result = Vec::new();
                for row in rows {
                    result.push(row?);
                }
                return Ok(result);
            }

            let mut conditions: Vec<String> = Vec::new();
            let mut params: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

            for e in &cues.entities {
                conditions.push("entities LIKE ?".to_string());
                params.push(Box::new(format!("%{e}%")));
            }

            for t in &cues.topics {
                conditions.push("topics LIKE ?".to_string());
                params.push(Box::new(format!("%{t}%")));
            }

            let sql = format!(
                "SELECT * FROM memories WHERE ({}) ORDER BY importance DESC, created_at DESC LIMIT ?",
                conditions.join(" OR ")
            );
            params.push(Box::new(limit as i64));

            let mut stmt = conn.prepare(&sql)?;
            let param_refs: Vec<&dyn rusqlite::types::ToSql> = params.iter().map(|p| p.as_ref()).collect();
            let rows = stmt.query_map(param_refs.as_slice(), crate::db::parse_memory_row)?;

            let mut result = Vec::new();
            for row in rows {
                result.push(row?);
            }
            Ok(result)
        })
    }
}

/// Fast cue extraction (regex only, <50ms).
fn extract_cues(text: &str) -> Cues {
    let mut entities = HashSet::new();
    let mut topics = HashSet::new();

    // Capitalized phrases (Entities)
    for cap in CAP_PATTERN.captures_iter(text) {
        if let Some(m) = cap.get(1) {
            if m.as_str().len() > 3 {
                entities.insert(m.as_str().to_string());
            }
        }
    }

    // #tags (Topics)
    for cap in TAG_PATTERN.captures_iter(text) {
        if let Some(m) = cap.get(1) {
            topics.insert(m.as_str().to_string());
        }
    }

    // @mentions (Entities)
    for cap in MENTION_PATTERN.captures_iter(text) {
        if let Some(m) = cap.get(1) {
            entities.insert(m.as_str().to_string());
        }
    }

    // Quoted strings (Entities/Specifics)
    for cap in QUOTE_PATTERN.captures_iter(text) {
        let val = cap.get(1).or_else(|| cap.get(2));
        if let Some(v) = val {
            if v.as_str().len() > 2 {
                entities.insert(v.as_str().to_string());
            }
        }
    }

    Cues {
        entities: entities.into_iter().collect(),
        topics: topics.into_iter().collect(),
    }
}
